exchange/exchange.go

```diff
@@ -637,6 +646,7 @@ func (e *exchange) getAllBids(
 	experiment *openrtb_ext.Experiment,
 	hookExecutor hookexecution.StageExecutor,
 	pbsRequestStartTime time.Time,
+	tmaxAdjustments *config.TmaxAdjustments,
```

Comment:
PR makes changes to pass tmaxAdjustments as pointer within different function. This results in additional tmaxAdjustments != nil check whenever we need to refer any field from tmaxAdjustments.
Not sure if we need to pass tmaxAdjustments as pointer or need to maintain same copy of tmaxAdjustments here

@VeronikaSolovei9 any thoughts on whether tmaxAdjustments should be or not passed as pointer?

Reply:
IMO, passing a pointer is more efficient memory wise than passing an entire struct since that will copy the values of fields in that struct.

Reply:
Not necessarily. There are optimizations in Go for passing arguments by value within the stack without needing to involve heap access. I recommend to keep with pass by value unless you can prove with benchmarks that it's better to use a pointer.

Reply:
Thanks @SyntaxNode !

Here are the benchmark results - 
```
ag476c@16612-p-mac exchange % go test -bench=. -benchmem -count 5 -run=^#
goos: darwin
goarch: amd64
pkg: github.com/prebid/prebid-server/exchange
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkTestUsingPtr-16       	 4693398	       242.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingPtr-16       	 4908163	       238.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingPtr-16       	 5311866	       236.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingPtr-16       	 5248773	       239.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingPtr-16       	 5184895	       246.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingStruct-16    	 1902420	       621.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingStruct-16    	 1953494	       565.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingStruct-16    	 1948306	       560.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingStruct-16    	 2106958	       601.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestUsingStruct-16    	 1988678	       597.0 ns/op	       0 B/op	       0 allocs/op
PASS
```

I tested it with small piece of code to build the stack for 100 iterations. Here is the code - 
```
func testBenchMarkUsingPtr(str *LargeStruct, n int) {
	for i := 0; i < n; i++ {
		if str != nil {
			testBenchMarkUsingPtr(str, 0)
		}
	}
}

func testBenchMarkUsingStruct(str LargeStruct, n int) {
	for i := 0; i < n; i++ {
		testBenchMarkUsingStruct(str, 0)
	}
}
```
I ran the tests couple of times on my local machine and got the similar results. 
From the above experiment, it looks like passing a pointer is much more efficient than to pass by value even with the nil check for pointer variable. The use of pointers allowed the code to perform a significantly higher number of operations compared to passing by value.

Reply:
@gargcreation1992, 

do `testBenchMarkUsingPtr(str, 0)`  and `testBenchMarkUsingStruct(str, 0)` calls getAllBids internally? IMO if we are using `number of operations` then it should be for actual function (getAllBids  in this case) and not of `testBenchMarkUsingPtr(str, 0)` and `testBenchMarkUsingStruct(str, 0)` 

also, benchmarking was to find whether passing struct as value uses more memory than passing pointer struct. Benchmark results are showing `0 B/op` and `0 allocs/op` for both cases

I found this article https://medium.com/a-journey-with-go/go-should-i-use-a-pointer-instead-of-a-copy-of-my-struct-44b43b104963 , which does benchmarks for passing struct as pointer vs value.  See if this helps

Reply:
@onkarvhanumante 
> do `testBenchMarkUsingPtr(str, 0)` and `testBenchMarkUsingStruct(str, 0)` calls getAllBids internally? IMO if we are using `number of operations` then it should be for actual function (getAllBids in this case) and not of `testBenchMarkUsingPtr(str, 0)` and `testBenchMarkUsingStruct(str, 0)`

No, both the test functions are not calling getAllBids internally. I think we wanted to find the general case where passing a pointer is more efficient vs passing a struct value. The test reflects the same. 
Benchmarking the getAllBids function alone may not provide accurate results since it involves other function calls within its execution. The objective here is to determine whether passing a pointer is generally more efficient than passing a struct. In the test above, I tried to build the stack and heap up for both the scenarios and capture the result.
 
> also, benchmarking was to find whether passing struct as value uses more memory than passing pointer struct. Benchmark results are showing `0 B/op` and `0 allocs/op` for both cases. So passing as pointer and pass by value are not affecting any memory. 
> 
> found this article https://medium.com/a-journey-with-go/go-should-i-use-a-pointer-instead-of-a-copy-of-my-struct-44b43b104963 , which does benchmarks for passing struct as pointer vs value. See if this helps

In the article above, there are two scenarios that has been discussed. First one is more about the struct initialisation. Second one is about passing the struct by value/pointer to a function. And in fact the result shared by above post reflects the similar results that we have got with the above benchmarking results. The passing by pointer one took less time than passing by value. 

In more general case, passing struct by pointer makes sense where function is mutating the struct fields in the function else we should pass by value. The concern here for optimisation is because we are dealing with billions of requests in a day and even slight optimisation makes a difference.
 
But I also realise that tmaxAdjustments struct is not much big struct. So, we could take this off PR and resolve this comment for now. Let me know thoughts.

Reply:
@onkarvhanumante and @SyntaxNode even though I'd normally agree with you about pointer usage, I think this looks like it might be a legit use case where using a pointer will likely be more performant. I normally consider the following as valid use cases for using a pointer:

- When you need to mutate the object
- When the object is extremely large in size. (This one is a little hard to reason for because there's no standard definition of large here. So for this one, benchmarking can help)
- When you need to be able to differentiate between the struct's zero value vs it being unset
- And lastly, when you're making intensive function calls either on that object or function calls where you're passing around that object.

I think the use case we have here for Tmax Adjustments, fits the last bucket. Tmax Adjustments is a host configuration object which is created once at startup and then we pass it around to each request so that we can calculate the bidder tmax for that request based on the Tmax Adjustments configuration. If we pass this object by value, we will be copying this object over and over again for each and every request vs if we pass it by pointer, we will be creating this object only once during startup and then just pass the reference to each request. In case of passing it by pointer, even though the initial overhead of allocating the object will be high, I think it will pay off over time.

This is also the second use case the article referenced by @onkarvhanumante above is talking about and the benchmarking shows that in such a case using a pointer is better than using a copy.

Reply:
thanks @mansinahar. Based on above I think we should use a pointer rather than using a copy.